### PR TITLE
Saving and Loading Stacked Ensemble Classifiers in ONNX format in Python

### LINKED ISSUE #5060: [Machine learning] Saving and Loading Stacked Ensemble Classifiers in ONNX format in Python
### Topic Suggestion

[Machine learning] Saving and Loading Stacked Ensemble Classifiers in ONNX format in Python

#### Proposed article introduction
Stacked ensemble models are learners that increase predictive performance over stand-alone learners by combining the results of two or several models and running them through a meta-learner. The stacked models are different (not a single type), unlike in bagging methods (just decision trees) and each of the model in the stack does not correct the predictions of the previous ones as it happens in boosting.

Saving this type of a model in ONNX format and using it in production can prove challenging as this scenario is not adequately documented.

ONNX is an open source format for deep and traditional machine learning developed by Microsoft that has a unified schema for saving models despite the library they were developed in. Launched in December 2017, it gives data scientists and Machine learning engineers a way to persist models without worrying about platform inconsistencies and library version deprecation.

#### Key takeaways 
In this article, the reader will learn how to:
  - install ONNX and onnxruntime
  - determine the ONNX input initialtypes 
  - serialize and save a stacked ensemble to ONNX format
  - load it to production using an Onnx runtime Inference session.

#### Article quality
Serializing stacked ensemble models in ONNX format and using it in production can prove challenging as this scenario is not adequately documented by the library maintainers.
The documentation is not friendly to beginners, thus a step-by-step article would be significantly helpful.

#### References
- [Scikit-Learn ONNX Introduction](http://onnx.ai/sklearn-onnx/introduction.html)
- [ONNX installation guide](https://pypi.org/project/onnx/)

### Conclusion
NA

### PR DESCRIPTION
**Pull request by iannjari**
Created:    /content/articles/save-and-load-stacked-ensembles-in-onnx/index.md
This article gives a tutorial on Saving and Loading Stacked Ensemble Classifiers in ONNX format in Python
Image:   /content/articles/save-and-load-stacked-ensembles-in-onnx/index.md

Added Author Profile:   /content/authors/ian-njari/index.md
Avatar:   /content/authors/ian-njari/avatar.jpg
This closes #5060

### CODE REVIEW

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
+# Saving and Loading Stacked Ensemble Classifiers in ONNX format in Python
+
+Stacked ensemble models are learners that increase predictive performance over stand-alone learners by combining the results of two or several models and running them through a meta-learner. The stacked models are different (not a single type), unlike in bagging methods (just decision trees) and each of the model in the stack does not correct the predictions of the previous ones as it happens in boosting. You can learn how to build one on [this](https://www.section.io/engineering-education/ensemble-learning-based-regression-model-using-python/) Sections article by Adhinga Fredrick.
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:3]:
```suggestion
Stacked ensemble models are learners that increase predictive performance over stand-alone learners by combining the results of two or several models and running them through a meta-learner. The stacked models are different (not a single type), unlike in bagging methods (just decision trees) and each of the model in the stack does not correct the predictions of the previous ones as it happens in boosting. You can learn how to build one by reading [this article by Adhinga Fredrick](/engineering-education/ensemble-learning-based-regression-model-using-python/).
```

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+
+Open Neural Network Exchange (ONNX) is an open source format for deep and traditional machine learning developed by Microsoft that has a unified schema for saving models despite the library they were developed in. Launched in December 2017, it gives data scientists and Machine learning engineers a way to persist models without worrying about platform inconsistencies and library version deprecation.
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:6]: Add link for ONNX

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+It is also key as a means to avoid vendor lockin since ONNX models can be deployed on any platform - not just where they were trained.
+Container-based methods of pushing models to production like Docker can also be by-passed altogether.
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:8]: Could you brief more about these 2 lines?

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+Saving this type of a model in ONNX format and using it in production can prove challenging as this scenario is not adequately documented.
+So how can stacked ensembles be serialized using ONNX?
+
+## Table of Contents
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:13]:
```suggestion
### Table of contents
```

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+  1. Preparing the environments
+  2. Importing and preparing the data
+  3. Building and evaluating the classifier
+  4. Serializing the model to ONNX format
+  5. Loading the model using ONNX runtime Inference Session
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:18]: Make these unordered lists using `-`, and indent them towards left.

Add hyperlinks if possible

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+### Prerequisites
+  1. Basic knowledge of Python
+  2. Machine learning model building, evaluation and validation in Scikit-Learn
+  3. Basic data manipulation skills
+  4. Python(with pip,numpy,pandas and Scikit-Learn) installed on your computer, or an online environment like Google Colab or Kaggle
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:24]: Use `-`

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+  - install ONNX and onnxruntime
+  - determine the ONNX input initialtypes 
+  - serializing and saving a stacked ensemble to ONNX format
+  - loading it to production using an Onnx runtime Inference session.
+
+### Setting up environments
+
+To install ONNX and onnxruntime on a local environment, run the following commands;
+
++ If using pip, on your terminal;
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:38]: Use `-` for bullet points

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+```
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:39]:
```suggestion
```bash
```

Comment (srishilesh) [reply content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:39]: Do this change for all the code snippets

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+!pip install onnx
+```
+and
+```
+!pip install onnxruntime
+```
+**Note:** Online editors like repl.it may fail to run the code below in full due to in-sufficient memory allocations.
+
+### Importing and preparing the data
+Let's start by importing pandas and the data
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:69]: Explain more about the dataset, and where are we getting this from?

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+```python
+import pandas as pd
+path='https://raw.githubusercontent.com/iannjari/datasets/main/diabetes.csv'
+df=pd.read_csv(path,engine='python')
+df
+```
+Output:
+
+![DataFrame](df1.png)
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:78]: Image path is incorrect. Please rectify.

Do this for all the images

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+*Image by author*
+
+We will seperate the data columns from the label/target, "Outcome";
+
+```python
+target_name = "Outcome"
+target = df[target_name]
+
+data = df.drop(columns=[target_name])
+data
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:88]: Use print(), since this would be a python code snippet and not a notebook script

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+```python
+from sklearn.model_selection import train_test_split
+x_train, x_test, y_train, y_test = train_test_split(
+    data,target, test_size=0.33, random_state=42)
+```
+
+### Training and evaluating the stacked classifier
+We shall employ a stack with a randomForest, kNN classifier, GradientBoosting classifier and a logistic regressor as a final model.
+Import packages;
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:100]: Add links that speak about these models and give a single line explanation of each

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+```python
+pipeline = make_pipeline(
+            StackingClassifier(estimators=[
+            ("rf",RandomForestClassifier(n_estimators=10,random_state=42)),
+            ("gb",GradientBoostingClassifier(n_estimators=10,random_state=42)),
+            ("knn",KNeighborsClassifier(n_neighbors=5))],final_estimator=LogisticRegression()))
+
+pipeline.fit(x_train,y_train)
+pipeline.score(x_test,y_test)
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:126]: print()

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+                      initial_types=
+                      initial_types)
+
+with open("stacked_clf.onnx", "wb") as f:
+    f.write(onx.SerializeToString())
+```
+The model is saved sucessfully.
+
+
+***Note:** If establishing the initial types is challenging, for example, the data has too many features, the ```to_onnx``` method handles this directly for you.
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:192]:
```suggestion
> NOTE: If establishing the initial types is challenging, for example, the data has too many features, the ```to_onnx``` method handles this directly for you.
```

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+You just need to pass the x test data (or one of it's column) as an argurment and ONNX extracts it automatically.
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:193]:
```suggestion
You just need to pass the `x_test` data (or one of it's column) as an argurment and ONNX extracts it automatically.
```

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+Happy ML-ing!
+
+### References
+[Tutorial on Building an Ensemble Learning Based Regression Model using Python](https://www.section.io/engineering-education/ensemble-learning-based-regression-model-using-python/), *Adhinga Fredrick, Sections Engineering*
+
+[ONNX Homepage](https://onnx.ai/)
+
+[ONNX sklearn documenation](http://onnx.ai/sklearn-onnx/introduction.html)
+
+[Common errors with onnxruntime](http://onnx.ai/sklearn-onnx/auto_examples/plot_errors_onnxruntime.html?highlight=errors)
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:253]: Make these as bulleted points

File: content/articles/save-and-load-stacked-ensembles-in-onnx/index.md [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+        0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0,
+        0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1,
+        1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
+        1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0], dtype=int64)]
+```        
+###  Conclusion
+In this tutorial, you learnt how to install ONNX and onnxruntime, determine ONNX input initialtypes, serializing and saving a stacked ensemble to ONNX format and finally loading it to production using an Onnx runtime Inference session.
+This model can now be served through a Web App framework like Streamlit and Dash, used in the backend (Django, Flask) or an API.
+In case of a hard nut to crack with ONNX you can raise an issue on [onnx's github](https://github.com/onnx/sklearn-onnx/issues)
+
```
Comment (srishilesh) [inline content/articles/save-and-load-stacked-ensembles-in-onnx/index.md:243]: Add an URL containing full working code. (It can be colab notebook or Github repo)

**Review by srishilesh — CHANGES_REQUESTED** [review]
@iannjari 

Welcome to EngEd program!

Great article. I've suggested a few changes. Have a look at them.

Let's wait for plagiarism check, before we start the next round of review.

Ping me when you're done with the changes.

Thanks.

**Comment by hectorkambow** [comment]
@iannjari 
 it seems this PR is still quite a bit aways from being completed. 
 
As a rule of thumb - please only submit articles (pull requests) with a complete and fully polished article - ready to be published. 
 
All revisions and edits should be completed on your own forked repo (as to not take up room in the queue). 
 
We will be closing this and please re-submit when it is ready to be published as a new PR (and you can add the duplicate label) and ping the reviewer and myself. 

Closing a PR does NOT mean the article can not be published (or fixed) it simply means that the edits, revisions, and fixes will happen outside of the queue.
 
👍 
cc @srishilesh

**Comment by srishilesh** [comment]
@iannjari 

You can create a new PR with all these suggested changes. Ping me once you're done.

Make sure that your new PR is ready to be published, and needs no extra revisions.

Thanks.
